firmware/shared/code/I2C.c

```diff
@@ -0,0 +1,244 @@
+	GPIO_PinAFConfig(GPIOF, GPIO_PinSource1, GPIO_AF_I2C2);
+
+
+	I2C_setup();
+
+	// NVIC Init
+	NVIC_InitTypeDef NVIC_Init_Struct;
+	NVIC_Init_Struct.NVIC_IRQChannel                   = I2C2_EV_IRQn;
+	NVIC_Init_Struct.NVIC_IRQChannelCmd                = ENABLE;
+	NVIC_Init_Struct.NVIC_IRQChannelPreemptionPriority = 5;
+	NVIC_Init_Struct.NVIC_IRQChannelSubPriority        = 0;
+	NVIC_Init(&NVIC_Init_Struct);
+
+	// Set priority for ISR and enable ISR
+	//NVIC_SetPriority(I2C2_EV_IRQn, 0);
+	NVIC_EnableIRQ(I2C2_EV_IRQn);
+}
+
+void I2C_setup(void)
+{
+	// I2C2 Init
+	RCC_APB1PeriphClockCmd(RCC_APB1Periph_I2C2, ENABLE);
+
+	// Reset I2C after a delay
+	// If you want to know why, it's because I2C won't work unless doing so.
+	int i = 0;
+	while (i < 20){
+		i++;
+	}
+	RCC_APB1PeriphResetCmd(RCC_APB1Periph_I2C2, ENABLE);
+	RCC_APB1PeriphResetCmd(RCC_APB1Periph_I2C2, DISABLE);
+
+	I2C_InitTypeDef I2C_Init_Struct;
+	I2C_Init_Struct.I2C_ClockSpeed          = 50000; // Any requirement for I2C clock speed?
+	I2C_Init_Struct.I2C_Ack                 = I2C_Ack_Enable;
+	I2C_Init_Struct.I2C_AcknowledgedAddress = I2C_AcknowledgedAddress_7bit;
+	I2C_Init_Struct.I2C_DutyCycle           = I2C_DutyCycle_2;
+	I2C_Init_Struct.I2C_Mode                = I2C_Mode_I2C;
+	I2C_Init_Struct.I2C_OwnAddress1         = 0x0;
+	I2C_Init(I2C2, &I2C_Init_Struct);
+
+	// Enable event interrupts, buffer interrupts, and error interrupt (for acknowledge failure)
+	I2C_ITConfig(I2C2, I2C_IT_EVT, ENABLE);
+	I2C_ITConfig(I2C2, I2C_IT_BUF, ENABLE);
+
+	// Enable I2C2
+	I2C_Cmd(I2C2, ENABLE);
+}
+
+bool I2C_send(I2C_channel_E channel, const uint8_t * data, const uint8_t length)
+{
+	if((data != NULL) && (length != 0) && (length <= _I2C_BUFFER_SIZE))
+	{
+		if ((data_length != 0) || (I2C_state != I2C_STATE_IDLE)) return false;
+		memcpy(I2C_send_data.Buffer, data, length);
+		I2C_state = I2C_STATE_SEND;
+		slave_address = I2C_channelToAddressMapping(channel);
+		data_length = length;
+		I2C_GenerateSTART(I2C2, ENABLE);
+		return true;
+	}
+	else return false;
+}
+
+bool I2C_receive(I2C_channel_E channel, const uint8_t * data, const uint8_t length)
+{
+	if((data != NULL) && (length != 0) && (length <= _I2C_BUFFER_SIZE))
+	{
+		if ((data_length != 0) || (I2C_state != I2C_STATE_IDLE)) return false;
+		memcpy(I2C_receive_data.Buffer, data, length);
+		I2C_state = I2C_STATE_RECEIVE;
+		slave_address = I2C_channelToAddressMapping(channel);
+		data_length = length;
+		I2C_GenerateSTART(I2C2, ENABLE);
+		return true;
+	}
+	else return false;
+}
+
+bool is_idle()
+{
+	if (I2C_state != I2C_STATE_IDLE) return false;
+	else return true;
+}
+
+// See stm32f413 programming manual P855-861
+void I2C2_EV_IRQHandler(void)
+{
+	// Check Start condition
+	if ((I2C2->SR1 & I2C_SR1_SB) == I2C_SR1_SB)
+	{
+		// LSB is the r/w bit
+		//I2C_Send7bitAddress(I2C2, (slave_address << I2C_OAR1_ADD0), (uint8_t)I2C_state);
+		uint8_t addr = (slave_address << I2C_OAR1_ADD0) + (uint8_t)I2C_state;
+		I2C2->DR = addr;

```

Comment:
Just do I2C2->DR = (slave_address << I2C_OAR1_ADD0) + (uint8_t)I2C_state; here instead of having the extra variable.

Reply:
I switched back to send7bitaddress function for now. I will check to make sure that function works consistently.